common/txmgr/confirmer.go

```diff
@@ -1218,6 +1225,37 @@ func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) Res
 	return nil
 }
 
+// Set latest finalized block number using the latest head
+func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) SetLatestFinalizedBlockNum(ctx context.Context, head types.Head[BLOCK_HASH]) {
+	if head.LatestFinalizedHead() != nil && head.LatestFinalizedHead().BlockNumber() != 0 {
+		ec.finalizedBlockNumMu.Lock()
+		ec.latestFinalizedBlockNum = head.LatestFinalizedHead().BlockNumber()
+		ec.finalizedBlockNumMu.Unlock()
+	}
+
+	ec.lggr.Debugf("set latest finalized block num: %d", ec.latestFinalizedBlockNum)
+}
+
+// Determines if a transaction is finalized by comparing its receipt's (if one exists) block num against the latest finalized block num
+// If the receipt's block num is less than or equal to the finalized block num, the transaction is considered finalized
+func (ec *Confirmer[CHAIN_ID, HEAD, ADDR, TX_HASH, BLOCK_HASH, R, SEQ, FEE]) CheckTransactionFinality(ctx context.Context, tx *txmgrtypes.Tx[CHAIN_ID, ADDR, TX_HASH, BLOCK_HASH, SEQ, FEE]) bool {
+	var receipt txmgrtypes.ChainReceipt[TX_HASH, BLOCK_HASH]
+	// Find tx receipt if one exists
+	for _, attempt := range tx.TxAttempts {
+		if len(attempt.Receipts) > 0 {
+			// Tx will only have one receipt
+			receipt = attempt.Receipts[0]
+			break
+		}
+	}
+	ec.finalizedBlockNumMu.RLock()
+	defer ec.finalizedBlockNumMu.RUnlock()
+	if receipt != nil && ec.latestFinalizedBlockNum != 0 && receipt.GetBlockNumber().Cmp(big.NewInt(ec.latestFinalizedBlockNum)) <= 0 {
```

Comment:
Using block numbers to identify the transaction's state seems unsafe. In some cases, we might return `finalized` state for transactions reorged from the chain. Example:
1. Confirmer sees receipt in block 10 and marks transaction as confirmed.
2. Confirmer fails to notice that the transaction is excluded from the chain. This might happen due to downtime (node goes down right after confirmation and starts after block 10 is finalized) or because HeadTracker fails to provide a long enough chain for the same number of blocks (this might happen if the block rate of the chain is faster than backfill).
3. When block 10 becomes finalized, we consider the transaction finalized even though the receipt is no longer present in the chain.

Reply:
Good point! Would adding a check to ensure the `receipt.GetBlockHash()` equals the `head.HashAtHeight(receipt.GetBlockNumber())` help mitigate this?

Reply:
Yes, that helps. But HeadTracker only provides chain up to latest finalized block + history depth, so if the transaction is older than that we'll return unconfirmed, while the transaction is finalized. 
So if the receipt's block is is older than earliest in chain, we need to make an additional check. One of the options is to make batch request to fetch latest finalized block and receipt's block by hash. If receipts block exists and it's older than latest finalized, we can return mark tx as finalized.

Reply:
I see. This might be an extreme edge case with the current expectations of how this is used. Don't think products will check status for txs that far back but I agree this is important to cover as use cases expand in the future or some abnormal downtime occurs. Also I'm ok with this option. Since we're not expecting a status query that far back in regular usage, I'm ok with making a batch RPC call in this scenario

Reply:
This might be a naive question but would I really need to fetch the latest finalized block too? Couldn't I rely on the latest finalized head from the head tracker for that? So the only call I have to make is the receipt block hash exists?

Reply:
Yes, you are right. It should be safe to do it the way you've described, but only with this [PR](https://github.com/smartcontractkit/chainlink/pull/13048) merged. Hopefully it'll happened next week.
Before that, there is no guarantee that RPC handling the request is not lagging behind on latest finalized block.

Reply:
Ah good to know. I'll mark that in the description. I think it makes sense to wait for this to go through so I'm not unnecessarily complicating the code.